base/unicode/properties.jl

```diff
@@ -0,0 +1,219 @@
+const global Pi = 16  # Punctuation, initial quote
+const global Pf = 17  # Punctuation, final quote
+const global Po = 18  # Punctuation, other
+const global Sm = 19  # Symbol, math
+const global Sc = 20  # Symbol, currency
+const global Sk = 21  # Symbol, modifier
+const global So = 22  # Symbol, other
+const global Zs = 23  # Separator, space
+const global Zl = 24  # Separator, line
+const global Zp = 25  # Separator, paragraph
+const global Cc = 26  # Other, control
+const global Cf = 27  # Other, format
+const global Cs = 28  # Other, surrogate
+const global Co = 29  # Other, private use
+
+abstract UnicodeProperty
+
+abstract CharCategory   <: UnicodeProperty
+
+abstract CatLetter      <: CharCategory
+abstract CatMark        <: CharCategory
+abstract CatNumber      <: CharCategory
+abstract CatPunctuation <: CharCategory
+abstract CatSymbol      <: CharCategory
+abstract CatSeparator   <: CharCategory
+abstract CatOther       <: CharCategory
+
+abstract CatUpper       <: CatLetter
+
+abstract LowerCase      <: CatLetter
+abstract UpperCase      <: CatUpper
+abstract TitleCase      <: CatUpper
+abstract ModifierLetter <: CatLetter
+abstract OtherLetter    <: CatLetter
+
+abstract NonSpacingMark <: CatMark
+abstract CombiningMark  <: CatMark
+abstract EnclosingMark  <: CatMark
+
+abstract DecimalDigit   <: CatNumber
+abstract NumericLetter  <: CatNumber
+abstract OtherNumber    <: CatNumber
+
+abstract ConnectorPunct <: CatPunctuation
+abstract DashPunct      <: CatPunctuation
+abstract OpenPunct      <: CatPunctuation
+abstract ClosePunct     <: CatPunctuation
+abstract BegQuotePunct  <: CatPunctuation
+abstract EndQuotePunct  <: CatPunctuation
+abstract OtherPunct     <: CatPunctuation
+
+abstract MathSymbol     <: CatSymbol
+abstract CurrencySymbol <: CatSymbol
+abstract ModifierSymbol <: CatSymbol
+abstract OtherSymbol    <: CatSymbol
+
+abstract SpaceSeparator <: CatSeparator
+abstract LineSeparator  <: CatSeparator
+abstract ParagraphSeparator <: CatSeparator
+
+abstract NotAssignedChar <: CatOther
+abstract ControlChar     <: CatOther
+abstract FormatChar      <: CatOther
+abstract SurrogateChar   <: CatOther
+abstract PrivateUseChar  <: CatOther
+
+const c2t = [NotAssignedChar, UpperCase, LowerCase, TitleCase, ModifierLetter, OtherLetter,
+             NonSpacingMark, CombiningMark, EnclosingMark,
+             DecimalDigit, NumericLetter, OtherNumber,
+             ConnectorPunct, DashPunct, OpenPunct, ClosePunct,
+             BegQuotePunct, EndQuotePunct, OtherPunct,
+             MathSymbol, CurrencySymbol, ModifierSymbol, OtherSymbol,
+             SpaceSeparator, LineSeparator, ParagraphSeparator,
+	     ControlChar, FormatChar, SurrogateChar, PrivateUseChar]
+
+############################################################################
+
+
+"""
+Return different Unicode properties for character
+"""
+function charprop end
+
+charprop(::Type{CharCategory}, c) = c2t[charprop(CharCategoryCode, c)+1]
```

Comment:
I don't understand what you're doing here. You want to use an enum internally, but return a type? After more thinking, I think returning an enum would be more natural. If we want to support aggregate categories, like `CatNumber`, we could define these as global constant tuples like `const CatNumber = (DecimalDigit, NumericalLetter, OtherNumber)`. Then one would write `charprop(CharCategory, c) in CatNumber`.

These values could all be put into a small module called `UnicodeCategory` to avoid polluting the global namespace.

Finally, I wonder whether it wouldn't be better to use the codes (like `Sk`), which are standardized and shorter. Not sure.